core/src/env3/call/builder.rs

```diff
@@ -0,0 +1,190 @@
+// Copyright 2018-2019 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::marker::PhantomData;
+
+use crate::env3::{
+    call::{
+        state,
+        CallData,
+        Selector,
+    },
+    EnvTypes,
+    Result,
+};
+
+/// Represents a return type.
+///
+/// Used as a marker type to differentiate at compile-time between invoke and evaluate.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
```

Comment:
There are a number of `Debug` derives in the PR. Are they required?

Reply:
I think they don't matter because dead code elimination will throw them out again if not needed.